### PR TITLE
Hardware-ware training with weight gradients, AnalogContext

### LINKED ISSUE #94: Enhance decay_weights in AnalogSGD
## Description and motivation

Currently there are some aspects of using `decay_weights` in `AnalogSGD` that are a bit different from the standard `SGD`. It would be great if we could refine the remaining aspects and revise the current documentation in order to highlight the differences and potential pitfalls to users, and streamline a bit its usage.

## Alternatives and other information

This is meant to be a small, "refinement" issue - not meaning to include bigger changes to the optimizer.

### LINKED ISSUE #211: Cleaner way of passing values from backward to update
## Description and motivation

Currently, we are passing the values needed for the `update()` by "attaching" them as attributes of the `layer.weights` tensor (during `backward()`). We should try to find a better way, perhaps using `model.buffers` instead.

### PR DESCRIPTION
**Pull request by maljoras**
## Related issues

makes foundation of #48 
closes #94  (any torch weight decay can now be used for inference tiles for hardware aware training)
closes #211 

The use of `weight` and `bias` as the proxy for the analog updater was awkward, since these were added as parameters and then removed again in `AnalogSgd`.  Especially for inference one might want to use the standard torch optimizers.  

## Description
Now inference tiles uses shared weights, which means that all standard torch optimizers can be used in this case in principle. This is because, the inference tiles now return the weight gradient in the backward pass. For analog training, the weight gradient is still directly applied in-memory as before.    

## Details

* A new `AnalogContext` is introduced which inherits from `Parameter`. This is the only proxy needed for the `AnalogSGD`. 
* In case of "shared_weights" (for inference only) weights are set to the torch memory and handled as a normal Parameter. Thus all SGD should work for inference tiles (including distributed training in principle).
* Regrouping is not necessary any more, because only an `AnalogContext` is inserted as parameter, and not the weights/bias of AnalogLinear etc. layers.
* weight / bias field is still used for syncing as before, but `state_dict` will not save these values, since they are not `Parameters` anymore. The weight is handled by the tile instead.   
* Also fixes a segfault in case of shared weights usage

### CODE REVIEW

**Review by diego-plan9 — APPROVED** [review]
Great addition Malte! So glad to finally be able to close #211 and for the new features overall.
